Ceph dashboard user management from the UI

ricardoasmarques opened the pull request:
Users with '**user**' scope permissions will be able to manage ceph dashboard users from the UI.

This PR includes both _frontend_ and _backend_ implementation.

![screenshot from 2018-06-28 10-30-59](https://user-images.githubusercontent.com/14297426/42030093-9bc53a72-7ac9-11e8-9cb9-0e6342513887.png)

![screenshot from 2018-06-28 10-44-39](https://user-images.githubusercontent.com/14297426/42030097-9ec78054-7ac9-11e8-976d-d44aec2429e3.png)

rjfd commented on src/pybind/mgr/dashboard/controllers/role.py:
```diff
@@ -0,0 +1,15 @@
+
+from ..services.access_control import ACCESS_CTRL_DB, SYSTEM_ROLES
+from . import ApiController, RESTController
+from ..security import Scope
```
nit: re-order the imports to:
```python
from . import ApiController, RESTController
from ..security import Scope
from ..services.access_control import ACCESS_CTRL_DB, SYSTEM_ROLES
```

rjfd commented on src/pybind/mgr/dashboard/controllers/user.py:
```diff
@@ -0,0 +1,97 @@
+from ..services.access_control import ACCESS_CTRL_DB, SYSTEM_ROLES
+from . import ApiController, RESTController
+from ..security import Scope
+from ..tools import Session
```
nit: re-order the commits

votdev commented:
I'd expect the form can't be submitted without selecting a role. Currently kicked out my admin after removing the 'adminstrator' role and pressing 'Submit'. :-\

rjfd commented on src/pybind/mgr/dashboard/controllers/user.py:
```diff
@@ -0,0 +1,97 @@
+            raise cherrypy.HTTPError(404)
+        ACCESS_CTRL_DB.save()
+
+    def set(self, username=None, password=None, name=None, email=None, roles=None):
```
the `username` parameter cannot be optional since it corresponds to the `RESOURCE_ID` of this REST controller. (I know the controller infrastructure should do these kind of validations, but we still haven't reached that point)

rjfd commented on src/pybind/mgr/dashboard/controllers/user.py:
```diff
@@ -0,0 +1,97 @@
+        except UserDoesNotExist:
+            raise cherrypy.HTTPError(404)
+        user_roles = []
+        if roles:
```
the code in this conditional block is the same as in the `create` method. you could extract this into a class method.

votdev commented:
Strange render issue.
![fireshot capture 39 - ceph - http___localhost_4200_ _users_add](https://user-images.githubusercontent.com/1897962/42037956-f1a6839a-7ae9-11e8-81c6-6bab1b92a818.png)

votdev commented:
Strange render issue in role dropdown.
![fireshot capture 40 - ceph - http___localhost_4200_ _users_add](https://user-images.githubusercontent.com/1897962/42038089-5a694cfa-7aea-11e8-9975-130ac7020a81.png)

rjfd commented:
@votdev @ricardoasmarques 
> I'd expect the form can't be submitted without selecting a role. Currently kicked out my admin after removing the 'adminstrator' role and pressing 'Update User'. :-\

I think there are two different issues:
The first one is: creating a user without any role associated is allowed by the CLI implementation, so I'm not against of also allowing it in the web UI.

The second one is: maybe we shouldn't allow a user to edit the roles of its own user.

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/select-badges/select-badges.component.spec.ts:
```diff
@@ -0,0 +1,26 @@
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
```
Please add tests for every functionality of the shared component.

ricardoasmarques replied:
@Devp00l I've added tests to the `cd-select-badges` component.

callithea commented on src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.html:
```diff
@@ -0,0 +1,190 @@
+            <cd-select-badges
+                [data]="userForm.controls.roles.value"
+                [options]="allRoles"
+                emptyMessage="There are no roles."></cd-select-badges>
```
I think "There are no roles." can be a bit misleading (no roles available at all).
Maybe it should be rephrased to something like "Currently no roles selected." or "No roles have been selected." ?

capri1989 commented:
Maybe I've missed it but what's about the documentation of the different roles? Is it already documented somewhere? Right now I have an idea but I'm not completely sure which roles I need in which circumstances. Any kind of help or redirect to an existing documentation would be helpful.

LenzGr commented:
> Maybe I've missed it but what's about the documentation of the different roles?

It's in the manual already: http://docs.ceph.com/docs/master/mgr/dashboard/#user-and-role-management

ricardoasmarques commented:
@capri1989 @LenzGr Maybe the description of each role (in the backend) should be improved, but I think this doesn't have to be part of this PR - https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/services/access_control.py#L88

ricardoasmarques commented:
@votdev @rjfd I've addressed all your comments.

With regards to the removal of own user permissions, I've added the following warning if logged-in user removes "user" read or update permissions for his own user:

![screenshot from 2018-07-02 11-56-37](https://user-images.githubusercontent.com/14297426/42161266-2686b14c-7df2-11e8-8b1a-d79e784a066b.png)

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts:
```diff
@@ -0,0 +1,145 @@
+
+  getRequest(): UserFormModel {
+    const userFormModel = new UserFormModel();
+    userFormModel.username = this.userForm.get('username').value;
```
Done

ricardoasmarques commented:
@Devp00l I've added frontend unit tests

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/shared/api/user.service.ts:
```diff
@@ -0,0 +1,32 @@
+    return this.http.get(`api/user/${username}`);
+  }
+
+  create(user: UserFormModel) {
```
As we have API tests for all other services it would be kind to maintain that even so it's not very fancy ;)

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/select-badges/select-badges.component.ts:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  private updateOptions() {
+    this.data.splice(0, this.data.length);
```
This function could be refactored. Here my diff:
```diff
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/shared/components/select-badges/select-badges.component.ts b/src/pybind/mgr/dashboard/frontend/src/app/shared/components/select-badges/select-badges.component.ts
index 01d0904..d6a626a 100644
--- a/src/pybind/mgr/dashboard/frontend/src/app/shared/components/select-badges/select-badges.component.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/shared/components/select-badges/select-badges.component.ts
@@ -14,7 +14,7 @@ export class SelectBadgesComponent implements OnChanges {
   constructor() {}
 
   ngOnChanges() {
-    if (this.options) {
+    if (this.data.length > 0) {
       this.options.forEach((option) => {
         if (this.data.indexOf(option.name) !== -1) {
           option.selected = true;
@@ -24,12 +24,7 @@ export class SelectBadgesComponent implements OnChanges {
   }
 
   private updateOptions() {
-    this.data.splice(0, this.data.length);
-    this.options.forEach((option: SelectBadgesOption) => {
-      if (option.selected) {
-        this.data.push(option.name);
-      }
-    });
+    this.data = this.options.filter(option => option.selected).map(option => option.name);
   }
 
   selectOption(option: SelectBadgesOption) {

```

ricardoasmarques replied:
`this.data = ...` will not work here because we cannot assign a new reference to a field that is an `@Input` of the component.

Devp00l replied:
Ok, than this function is fine with me, but could you change line 17 as the if clause will always resolve to true unless you give it a `undefined` input. IMO `this.data.length > 0` would make more sense there.

ricardoasmarques commented:
@Devp00l I've addressed your comments

LenzGr commented:
LGTM. One (minor) nit: when creating and editing a user's details, I'd suggest to skip the "show password" icons when moving from one input field to the next using the Tab key. Am I right that one currently can only assign existing roles to a user? Do you plan to provide roles management in a separate PR?

ricardoasmarques commented:
@LenzGr Yes, role management will be submitted in a separate PR. It's currently a work in progress - https://tracker.ceph.com/issues/24447

LenzGr commented:
Another observation: with multiple user accounts, we need a way to display the name of the currently logged in user. I created [#24822](https://tracker.ceph.com/issues/24822) on the tracker for that.

LenzGr commented:
If I were to make changes to my own account, I need to log out and log in again in order for the changes to take full effect, correct? Would it make sense to display a corresponding notification?

ricardoasmarques commented:
> If I were to make changes to my own account, I need to log out and log in again in order for the changes to take full effect, correct? Would it make sense to display a corresponding notification?

@LenzGr I've addressed your comment.
If current logged in user roles have been changes, user will be automaticaly logged out:

![screenshot_20180721_113132](https://user-images.githubusercontent.com/14297426/43034867-6e5527ee-8cdc-11e8-8b6d-fac3ef4175bc.png)

ricardoasmarques commented:
>  ... when creating and editing a user's details, I'd suggest to skip the "show password" icons when moving from one input field to the next using the Tab key

@LenzGr done

ricardoasmarques commented:
All comments on this PR have been addressed.

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.spec.ts:
```diff
@@ -0,0 +1,243 @@
+      SharedModule
+    ],
+    declarations: [UserFormComponent, FakeComponent]
+  });
```
Change it to `}, true);` as the `edit mode` tests will fail, except the first, without a reset of all loaded modules.

ricardoasmarques replied:
Done

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts:
```diff
@@ -0,0 +1,231 @@
+      if (roles.indexOf(role.name) !== -1 && role.scopes_permissions['user']) {
+        const userPermissions = role.scopes_permissions['user'];
+        return ['read', 'update'].every((permission) => {
+          return userPermissions.indexOf(permission) !== -1;
```
`return userPermission[permission]` would be as sufficient, as it would return `undefined` if not there which would resolve to `false`.

ricardoasmarques replied:
Note that `userPermission` is an array, so your suggestion does not work here, because

````
['read', 'update'].every((permission) => { 
  return ["create", "delete", "read", "update"].indexOf(permission) !== -1 
});
````
returns `true`, while:

````
['read', 'update'].every((permission) => { 
  return ["create", "delete", "read", "update"][permission]
});
````
returns `false`.

Devp00l commented on src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-form/user-form.component.ts:
```diff
@@ -0,0 +1,231 @@
+
+  private isUserRemovingNeededRolePermissions(): boolean {
+    const isCurrentUser = this.isCurrentUser();
+    return isCurrentUser && !this.hasUserReadUpdatePermissions(this.userForm.get('roles').value);
```
There are a view `this.userForm.get(xyz).value` in this component, which you can now replace with `this.userForm.getValue(xyz)`

ricardoasmarques replied:
From `CdFormGroup.getValue` documentation:

>  Get the value of a control if it has none it will return false

Note that `hasUserReadUpdatePermissions` is expecting an array, that's why I would prefer to not user `CdFormGroup.getValue` for production code.

IMO we should change the implementation of `CdFormGroup.getValue` in the future, to not transform `none` value into `false` value.

ricardoasmarques commented:
> Ok, than this function is fine with me, but could you change line 17 as the if clause will always resolve to true unless you give it a undefined input. IMO this.data.length > 0 would make more sense there.

@Devp00l yes, the input can be undefined if you are waiting for a response from server to populate the options. I'd prefer to keep this check (in line 17) so that the parent component don't have to initialize the options with an empty array just to prevent an error in the `select-badges.component`. With this check, the component is more flexible and easy to use.

ricardoasmarques commented:
@Devp00l I've addressed your comments.

ricardoasmarques commented:
@Devp00l  I've addressed your comments.

ricardoasmarques commented:
@LenzGr This PR is now rebased and passed all checks.
